statsmodels/base/distributed_estimation.py

```diff
@@ -0,0 +1,698 @@
+    ----------
+    mod : statsmodels model class instance
+        The model for the current partition.
+    params : array-like
+        The estimated coefficients for the current partition.
+    hess_kwds : dict-like or None
+        Keyword arguments for the hessian function.
+
+    Returns
+    -------
+    An array-like object, updated design matrix, same dimension
+    as mod.exog
+    """
+
+    # TODO correctly handle hessian_obs
+    rhess = np.sqrt(mod.hessian_obs(np.asarray(params), **hess_kwds))
+    return rhess[:, None] * mod.exog
+
+
+def _calc_nodewise_row(wexog, idx, alpha):
+    """calculates the nodewise_row values for the idxth variable, used to
+    estimate approx_inv_cov.
+
+    Parameters
+    ----------
+    wexog : array-like
+        The weighted design matrix for the current partition.
+    idx : scalar
+        Index of the current variable.
+    alpha : scalar or array-like
+        The penalty weight.  If a scalar, the same penalty weight
+        applies to all variables in the model.  If a vector, it
+        must have the same length as `params`, and contains a
+        penalty weight for each coefficient.
+
+    Returns
+    -------
+    An array-like object of length p-1
+
+    Notes
+    -----
+
+    nodewise_row_i = arg min 1/(2n) ||wexog_i - wexog_-i gamma||_2^2
+                             + alpha ||gamma||_1
+    """
+
+    p = wexog.shape[1]
+    # handle array alphas
+    if not np.isscalar(alpha):
+        alpha = alpha[ind]
+
+    ind = list(range(p))
+    ind.pop(idx)
+
+    tmod = OLS(wexog[:, idx], wexog[:, ind])
+
+    nodewise_row = tmod.fit_regularized(alpha=alpha).params
+
+    return nodewise_row
+
+
+def _calc_nodewise_weight(wexog, nodewise_row, idx, alpha):
+    """calculates the nodewise_weightvalue for the idxth variable, used to
+    estimate approx_inv_cov.
+
+    Parameters
+    ----------
+    wexog : array-like
+        The weighted design matrix for the current partition.
+    nodewise_row : array-like
+        The nodewise_row values for the current variable.
+    idx : scalar
+        Index of the current variable
+    alpha : scalar or array-like
+        The penalty weight.  If a scalar, the same penalty weight
+        applies to all variables in the model.  If a vector, it
+        must have the same length as `params`, and contains a
+        penalty weight for each coefficient.
+
+    Returns
+    -------
+    A scalar
+
+    Notes
+    -----
+
+    nodewise_weight_i = sqrt(1/n ||wexog,i - wexog_-i nodewise_row||_2^2
+                             + alpha ||nodewise_row||_1)
+    """
+
+    n, p = wexog.shape
+    # handle array alphas
+    if not np.isscalar(alpha):
+        alpha = alpha[ind]
+
+    ind = list(range(p))
+    ind.pop(idx)
+
+    d = np.linalg.norm(wexog[:, idx] - wexog[:, ind].dot(nodewise_row))**2
+    d = np.sqrt(d / n + alpha * np.linalg.norm(nodewise_row, 1))
+    return d
+
+
+def _calc_approx_inv_cov(nodewise_row_l, nodewise_weight_l, p):
```

Comment:
If the value of `p` can be deduced from other args better not to include it as an argument.